### PR TITLE
here we go at last

### PR DESCRIPTION
**Pull request by ghost**
this should be the fix for it all. Is there somewhere I can read up on git hub so i dont do this again

### CODE REVIEW

**Comment by joolswills** [comment]
I asked you not to open another PR and to use the issue tracker. Is there a reason you are ignoring that? There are many sites with tutorials and documentation on using git and github. Can be found easily via a search engine.

**Comment by ghost** [comment]
sorry i didnt see that message its done now sorry

**Comment by joolswills** [comment]
I need more information as to what the PR addresses - we generate a daphne input configuration from emulationstation if daphne is installed and a joypad reconfigured. I am not aware of any issues with it, and tested it when the input configuration was created.

Also if the changes could affect existing controller configs etc.

More detail needed. thanks.

**Comment by ghost** [comment]
Ill try be quick in this explaining this

		case SDL_JOYBUTTONDOWN:
			reset_idle(); // added by JFA for -idleexit

			// added by Russ
			// loop through buttons and look for a press
			for (i = 0; i < (sizeof(joystick_buttons_map) / sizeof(int)); i++) {
				if (event->jbutton.button == i) {
					input_enable((Uint8) joystick_buttons_map[i]);
					break;
				}
			}
my code change just removes loop  thats not needed.  The functions do the exact same thing




// button mapping for gp2x
int joystick_buttons_map[] =
{
0 , //KEY_UP
0 , //KEY_LEFT
1 , //KEY_DOWN
2 , //KEY_RIGHT
2 , //KEY_START1
2 , //KEY_START2
3 , //KEY_BUTTON1
0 , //KEY_BUTTON2
4 , //KEY_BUTTON3
9 , //KEY_COIN1
18 , //KEY_COIN2
19 , //KEY_SKILL1
6 , //KEY_SKILL2
7 , //KEY_SKILL3
8 , //KEY_SERVICE
20 , //KEY_TEST
6 , //KEY_RESET
6 , //KEY_SCREENSHOT
-1, //KEY_QUIT
-1, //KEY_PAUSE 
-1, //KEY_CONSOL
-1 //KEY_TILT
};
#endif

this is the problem these hard buttons are set by default they shouldnt be defaults like like this thats what messes the whole thing up on certain conditions. changing this alone to -1 will defaults all buttons to not set.



changing else if (event->jaxis.axis == 0)  fixed the analog joystick not working for left and right id 


										if (strcasecmp(key_name.c_str(), g_key_names[i])==0)
										{
											// if -1 then no mapping necessary, just use default, if any
											if ( val1 >= -0  ) g_key_defs[i][0] = val1;							
											if ( val2 >= -0  ) g_key_defs[i][1] = val2;
											if ( val3 >= -0  )  
											{
												joystick_buttons_map[val3] = i;
												printf("mapping joystick button:%d to %s\n",val3,g_key_names[i] );
											}												
											else 
											{
												printf("using hard code button:%d to , %s \n",joystick_buttons_map[i] ,g_key_names[i]);
											}
											found_match = true;
																						
											break;
										}	
									}
					
this just changes the joystick button numbers to what they a should and use -1 in the ini to instead of 0 to use defaults the lines above this code that read
									if( strstr( sval1.c_str(), "-")) val1 =-abs(val1);
									if( strstr( sval2.c_str(), "-")) val2 =-abs(val2);
									if( strstr( sval3.c_str(), "-")) val3 =-abs(val3);

i test for a - in the string the reason it was like this is because atoi doesnt support negative numbers

the original code subtracts one from the value when setting it up so its identical in function to whats its setting for the emulator at runtime the only change is in the ini file

there is a little mistake in this code as well

printf("using hard code button:%d to , %s \n",joystick_buttons_moystick_buttons_map[i] ,g_key_names[i]);

should read 

printf("using hard code button:%d to , %s \n",joystick_buttons_map[i] ,g_key_names[joystick_buttons_map[i]]);

thats just showing information though of whats happening when you dont set a key up for a joystick button












 






technicallyl the only changes that are need is the  joystick_buttons_map[]  initailised to all -1 for not set. 
and the axis == 0 for analogue joystick fixes

I just thought it would be nice to not have to add a number onto the joystick button number and being able to use all -1 to make the ini file more logical if thats a problem its easily fixed just dont apply that fix i didnt change any function of the code or variables


setting all keys to 0 like this in the original code 
[KEYBOARD]
KEY_UP =         273 114 0
KEY_LEFT =       276 100 0
KEY_DOWN =       274 102 0
KEY_RIGHT =      275 103 0
KEY_START1 =     049 000 0
KEY_START2 =     050 000 0
KEY_BUTTON1 =    306 097 0
KEY_BUTTON2 =    308 115 0
KEY_BUTTON3 =    032 113 0
KEY_COIN1 =      053 000 0
KEY_COIN2 =      054 000 0
KEY_SKILL1 =     304 119 0
KEY_SKILL2 =     122 105 0
KEY_SKILL3 =     120 107 0
KEY_SERVICE =    057 000 0
KEY_TEST =       283 000 0
KEY_RESET =      284 000 0
KEY_SCREENSHOT = 293 000 0
KEY_QUIT =       027 113 0
KEY_PAUSE =       0   0  0
KEY_CONSOLE =     0   0  0
KEY_TILT =        0   0  0

will leave you with the following defaults set to hard coded unless changed to another number

The way  to fix this the way it is now is you would have to set every button your not going to use  to something else in case it has a hard code in it .  a in 0 dainput  would just use that default hard code.



int joystick_buttons_map[] =
{
0 , //KEY_UP
0 , //KEY_LEFT
1 , //KEY_DOWN
2 , //KEY_RIGHT
2 , //KEY_START1
2 , //KEY_START2
3 , //KEY_BUTTON1
0 , //KEY_BUTTON2
4 , //KEY_BUTTON3
9 , //KEY_COIN1
18 , //KEY_COIN2
19 , //KEY_SKILL1
6 , //KEY_SKILL2
7 , //KEY_SKILL3
8 , //KEY_SERVICE
20 , //KEY_TEST
6 , //KEY_RESET
6 , //KEY_SCREENSHOT
-1, //KEY_QUIT
-1, //KEY_PAUSE 
-1, //KEY_CONSOL
-1 //KEY_TILT
};

**Comment by ghost** [comment]
ok updated the code it wont have any effect on the scripts you currently have "the default of add +1 to the button number" .

File: src/io/input.cpp [hunk 1]
```diff
@@ -103,46 +103,65 @@ int g_key_defs[SWITCH_COUNT][2] =
...
+6 , //KEY_SKILL2
+7 , //KEY_SKILL3
+8 , //KEY_SERVICE
+20 , //KEY_TEST
+6 , //KEY_RESET
+6 , //KEY_SCREENSHOT
+-1, //KEY_QUIT
+-1, //KEY_PAUSE 
+-1, //KEY_CONSOL
+-1 //KEY_TILT
```
Comment (joolswills) [inline src/io/input.cpp:160]: why is the gp2x code changed here ?

File: src/io/input.cpp [hunk 2]
```diff
@@ -210,24 +229,36 @@ void CFG_Keys()
...
+									if( strstr( sval1.c_str(), "-")) val1 =-abs(val1);
+									if( strstr( sval2.c_str(), "-")) val2 =-abs(val2);
+									if( strstr( sval3.c_str(), "-")) val3 =-abs(val3);
 									corrupt_file = false;	// looks like we're good
 
 									bool found_match = false;
 									for (int i = 0; i < SWITCH_COUNT; i++)
 									{
 										// if we can match up a key name (see list above) ...
+
```
Comment (joolswills) [inline src/io/input.cpp:242]: trailing whitespace

File: src/io/input.cpp [hunk 3]
```diff
@@ -210,24 +229,36 @@ void CFG_Keys()
...
+											}												
+											else 
+											{
+												printf("using hard code button:%d to , %s \n",joystick_buttons_map[i] ,g_key_names[i]);
+											}
 											found_match = true;
+																						
 											break;
-										}
+										}
```
Comment (joolswills) [inline src/io/input.cpp:260]: trailing whitespace.

**Comment by joolswills** [comment]
Have you tested it with the automatic input file generation we use in retropie ?

**Comment by ghost** [comment]
it will work the way it worked before. All i changed was the values to what they should be. The problem wasnt that pary of the code. I set my buttons up manually dont need any scripts to do i for me what the config says is what it will use

**Comment by joolswills** [comment]
Well, retropie uses the script so it needs to be tested with that. As of now, I am still not completely clear on what this resolves (despite your explanation) as I had no problem with the input controls before. Sorry.

**Comment by ghost** [comment]
just leave it if you want the analog joystick is broken and the default buttons need unset

**Comment by joolswills** [comment]
That's not really helpful. Unfortunately I am struggling with your explanation so I will leave it then. The PR would need work and cleanup anyway.

**Comment by ghost** [comment]
well the problem isint when you set input its when you dont bad things can happen the code explains what going on in itself

**Comment by ghost** [comment]
https://retropie.org.uk/forum/topic/16464/daphne-arcade-joystick-orientation-wrong/63

**Pull request closed without merging**